### PR TITLE
chore(ci): stash/unstash build artifacts

### LINKED ISSUE #5851: Reuse build artifacts on different build agents
**Description**

Since #5717 the IT stage of the ci pipeline is executed on another build agent. However, since build artifacts are not shared by default, the build artifacts had to be build a second time on the agent running the IT stage. To speed up the ci pipeline, we should reuse build artifacts when using a different build agent.

### PR DESCRIPTION
**Pull request by korthout**
## Description

This PR uses Jenkins's stash and unstash to reuse build artifacts between different nodes. This way the `IT (Java)` stage (running on another node) does not have rebuild the same maven artifacts.

## Related issues

closes #5851

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: .ci/scripts/distribution/it-java.sh
Comment (korthout) [inline .ci/scripts/distribution/it-java.sh:8]: Only the jars build by us in the `Build (Java)` stage are stashed and unstashed, we still need to download all external dependencies from nexus so this can no longer be offline

Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:8]: Can't we just stash the whole `.m2` folder?

**Review by npepinpe — COMMENTED** [review]
:tada: 

Some questions:

1. We stash the m2 folder and the Java artifacts - the integrations test don't need the Go ones, right?
1. I guess we still need to run `prepare.sh` in the other agent to install some required libraries? Or is this just for Go maybe?

File: .ci/scripts/distribution/it-java.sh
Comment (cmur2) [reply .ci/scripts/distribution/it-java.sh:8]: Please don't, that will depending on the size put severe load on the Jenkins central server as per https://www.jenkins.io/doc/pipeline/steps/workflow-basic-steps/#stash-stash-some-files-to-be-used-later-in-the-build:
> This is because stashed files are archived in a compressed TAR, and with large files this demands considerable on-master resources, particularly CPU time. There's not a hard stash size limit, but between 5-100 MB you should probably consider alternatives.

Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:8]: Hm, I suggested this because Optimize does it I think ;) What would you propose? Re-downloading all the artifacts is non negligible, and sometimes may cause builds to fail (which is one of the reason we decided to go offline after the initial pull in the past).

Comment (cmur2) [reply .ci/scripts/distribution/it-java.sh:8]: With stashing the artifacts produced by the `Build (Java)` stage we're already a bit over the limits suggested by the documentation. To preserve the whole `.m2` folder a completely different approach needs to be taken, e.g. using a GCloud storage bucket with subfolders per branch and build number. This in turn involves having the GCloud SDK/`gsutil` tool available to the build and proper authentication etc.

**Comment by cmur2** [comment]
>     2. I guess we still need to run `prepare.sh` in the other agent to install some required libraries? Or is this just for Go maybe?

The `prepare.sh` seems to be intended to run on every new Maven container on every fresh agent to also adjust file permissions etc. This can either be handled by a function extracting this logic into a reusable code piece, or, a new Docker image derived from your current Maven build image.

File: .ci/scripts/distribution/it-java.sh
Comment (cmur2) [reply .ci/scripts/distribution/it-java.sh:8]: Optimize caches also not everything but just org.camunda Maven group artifacts I think.

Re-downloading is quite fine because we employ Maven caches located in the same K8s cluster as the CI jobs. They are referred using cluster-local URLs so also quite isolated and stable.

**Comment by korthout** [comment]
> 1. We stash the m2 folder and the Java artifacts - the integrations test don't need the Go ones, right?

I guess not, but I'm not sure, should I exclude those explicitly?

> 2. I guess we still need to run prepare.sh in the other agent to install some required libraries? Or is this just for Go maybe?

prepare.sh installs libatomic1 which I guess should be available, but again I'm not sure.

File: .ci/scripts/distribution/it-java.sh
Comment (korthout) [reply .ci/scripts/distribution/it-java.sh:8]: @npepinpe do you see any other issues with not using `--offline`?

Comment (npepinpe) [reply .ci/scripts/distribution/it-java.sh:8]: I honestly can't remember, maybe @menski can?

File: Jenkinsfile
Comment (cmur2) [inline Jenkinsfile:77]: This change will be supported by https://github.com/camunda/infra-core/pull/1729 but I'm holding back merging https://github.com/camunda/infra-core/pull/1729 before it becomes clear that this PR get's green-lit.

**Review by cmur2 — APPROVED** [review]

**Review by npepinpe — APPROVED** [review]
Looks good to me in general, let's check Monday with menski if he remembers why we decided to take builds offline before merging :+1:

**Comment by npepinpe** [comment]
Checked in with menski, the reason for going offline was that building and skipping tests would cause us to skip downloading test dependencies, which means that in our parallel test steps artifacts might be downloaded in parallel, which sometimes caused problems and our tests to explode.

**Comment by korthout** [comment]
> artifacts might be downloaded in parallel

I think that can be resolved relatively easily. We could consider to extract building the test classes (i.e. `mvn test-compile`) and with that downloading the test dependencies from executing the tests. For example, moving it as a step at the end of the Build stage or by moving it as a step at the start of the IT stage.

**Comment by npepinpe** [comment]
I'll defer to @cmur2 here :slightly_smiling_face:

**Comment by cmur2** [comment]
> > artifacts might be downloaded in parallel
> 
> I think that can be resolved relatively easily. We could consider to extract building the test classes (i.e. `mvn test-compile`) and with that downloading the test dependencies from executing the tests. For example, moving it as a step at the end of the Build stage or by moving it as a step at the start of the IT stage.

I think @korthout suggestion makes a lot of sense, especially when done at the start of the IT stage. Could probably made part of the `it-java.sh` script to invoke `mvn` twice.

The alternative is to reject this PR which will create big overhead for setting up a different implementation to share artifacts between agents. In November I will have to shift my DRI attention to the Consulting team (I'm DRI of both) so it might delay the whole modularization of the pipeline.

**Comment by npepinpe** [comment]
> The alternative is to reject this PR which will create big overhead for setting up a different implementation to share artifacts between agents

Do you mean this PR will create overhead, or rejecting it will create overhead?

**Comment by cmur2** [comment]
> > The alternative is to reject this PR which will create big overhead for setting up a different implementation to share artifacts between agents
> 
> Do you mean this PR will create overhead, or rejecting it will create overhead?

I mean rejecting the approach this PR proposes (using stash/unstash) will create overhead.

So I'm all in favor of finding a solution to adjust this PR to the parallel artifact download issue.

**Comment by npepinpe** [comment]
Let's go with that then - I'm guessing building the test classes will still need to be done in all agents.

**Comment by npepinpe** [comment]
Anything I can help with here?

**Comment by cmur2** [comment]
> Anything I can help with here?

I'd really appreciate if @korthout (once he is relieved from his medic duties which I understand can take a lot of time) could take the lead in looking into the needed modification as I'm not too familiar with the Maven setup of this project. I'll happily provide input if necessary.

**Comment by npepinpe** [comment]
@korthout Let me know if you're too busy at the moment for this - I'd like to move this forward as I'm fixing the upgrade tests and reenabling them, and I'm running into some issues such as the dind sibling running of IPv4 addresses :smile: So I'd like to replicate this and split away the upgrade tests as well, but I'd wait for this to be done so we can do it properly.

All this to say, let me know and I'll find time or someone to do it :+1:

**Comment by korthout** [comment]
@npepinpe if tomorrow is fine, I can look at it then :) Shouldn't take too much time to finish this. If you need it sooner, then please defer

**Comment by korthout** [comment]
@npepinpe @cmur2 I first tried `mvn test-compile`, but that of course did not download the plugins.

We should actually use `dependency:go-offline` to download all dependencies. However, there are some known issues with it. I thought that using the latest version (i.e. `org.apache.maven.plugins:maven-dependency-plugin:3.1.2:go-offline`) would resolve these, but there is still at least 1 plugin that appears to not be available in the offline mode. *

There is an ongoing jira issue https://issues.apache.org/jira/browse/MDEP-82 that discusses the situation. In it a workaround is mentioned in the form of: https://github.com/qaware/go-offline-maven-plugin. How do you feel about trying this?

*: it claims that `org.codehaus.plexus:plexus-utils:jar:1.1` has not been downloaded, but the logs clearly show that it did download it.

**Comment by npepinpe** [comment]
Let's try it :+1:

**Comment by korthout** [comment]
@npepinpe @cmur2 Using the `go-offline-maven-plugin` and defining the dynamic dependencies explicitly, this now works correctly. I've also moved the execution of the plugin to a maven profile instead of calling it explicitly, which is a lot more safe because it makes sure that the versions defined in the pom are always used.

Please have another look at these changes.

**Review by cmur2 — APPROVED** [review]

File: Jenkinsfile
Comment (korthout) [reply Jenkinsfile:77]: @cmur2 should that change be made before this is merged?

Comment (cmur2) [reply Jenkinsfile:77]: The order here is arbitrary as the change is already applied in https://ci.zeebe.camunda.cloud/configfiles/editConfig?id=maven-nexus-settings-zeebe-local-repo so my PR https://github.com/camunda/infra-core/pull/1729 can go before or after, it merely makes official configuration as code what is applied in prod.

**Review by npepinpe — APPROVED** [review]
:+1: 

Thanks for this!

**Comment by korthout** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0543186b28f14b5eed2faacc82d66bacc38d6d2a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0543186b28f14b5eed2faacc82d66bacc38d6d2a
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by korthout** [comment]
@npepinpe as discussed, please have another look. (also note, I've added the backport labels).

**Review by npepinpe — APPROVED** [review]
Thanks Nico!

**Comment by npepinpe** [comment]
Seems like it failed to resolve some dependencies in the end :disappointed:

**Comment by korthout** [comment]
flaky test

bors retry

**Comment by npepinpe** [comment]
@korthout [reacting faster than the flash](https://www.youtube.com/watch?v=LfmrHTdXgK4)

**Comment by korthout** [comment]
exit code 4 is conflict. I'll backport manually

**Pull request merged**
